File: Python/43.Multiply Strings.py
#两个数相乘的结果(输入为两个字符串类型的数，输出为字符串)，要求数可以任意长
class Solution(object):
    def multiply(self, num1, num2):
        if num1=='0' or num2=='0':
            return '0'

        nu1 ,nu2 = num1[::-1] , num2[::-1]
        tmp , result = [] , ''
        for i,ch2 in enumerate(nu2):
            flag , t = 0 , ''
            t += '0'*i
            for ch1 in nu1:
                s = int(ch1)*int(ch2)+flag
                flag = s//10
                t += str(s%10)
            if flag!=0:
                t+=str(flag)

            tmp.append(t)
        #print(tmp)
        flag , l , i = 0 , len(t) , 0
        while i<l or flag!=0:
            s = flag
            for ch in tmp:
                if i<len(ch):
                    s += int(ch[i])

            result += str(s%10)
            flag = s//10
            i += 1
        return result[::-1]

s = Solution()
print(s.multiply('576','120'))

File: C++/String/43.Multiply Strings.cpp
/*
两个字符串代表两个非负整数，求两个数的积。
*/

#include <iostream>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
	string multiply(string num1, string num2) {
		vector<string>	strvec;
		string result = "";
		for (int i = num1.size() - 1; i >= 0; --i)
		{
			string temp = "";
			int flag = 0, multi;
			for (int j = num2.size() - 1; j >= 0; --j)
			{
				multi = (num1[i] - '0')*(num2[j] - '0') + flag;
				temp += char(multi % 10 + '0');
				flag = multi / 10;
			}
			temp = (flag == 0) ? temp : (temp + char(flag + '0'));
			strvec.push_back(temp);
		}
		int flag = 0, sum, n = strvec.size() + strvec.back().length() - 1;
		for (int i = 0; i < n; ++i)
		{
			sum = flag;
			for (int j = 0; j <= i && j<strvec.size(); ++j)
			{
				if (i - j<strvec[j].size())
					sum += (strvec[j][i - j] - '0');
			}
			result = char(sum % 10 + '0') + result;
			flag = sum / 10;
		}
		//最后一个进位
		if(flag != 0) result = char(flag + '0')+result;
		//去掉前面的0
		int pos = result.find_first_not_of('0');
		return (pos==-1)? "0" : result.substr(pos);
	}
};
int main()
{
	Solution s;
	cout<< s.multiply("15", "112")<<endl;
}

